### PR TITLE
Optimize out empty ifs

### LINKED ISSUE #9301: Optimize out empty If statements in LIR
Presently Logstash allows the creation of pipelines with empty `if` statements, i.e. a condition is provided and no plugins or subsequent `if` statements are specified within its scope.

```
input {
  stdin { }
}

filter {
  if [foo] in [bar] {

  }
}

output {
  stdout { }
}
```

I propose that we optimize out empty `if`s in LIR. This shouldn't impact the execution of pipelines and it will simplify some LS UI tasks we're working on.

### PR DESCRIPTION
**Pull request by ycombinator**
Resolves #9301.

If an if statement has no true or false sub-statements, there is no point including the if statement in the LIR graph. So instead of constructing a sub-graph for the if statement — which would include an empty if vertex and nothing else — and returning it, we return an empty graph object.

### CODE REVIEW

**Comment by ycombinator** [comment]
@andrewvc I realize this implementation isn't exactly what you had in mind (i.e. to create an `optimize()` method on `Graph` that would optimize away empty `if`s).

My thinking behind the implementation in this PR was: why create empty `if`s in the graph at all if we can detect them when converting from imperative -> IR? This way users could still specify empty `if`s in their imperative config but we'd optimize them away when creating the graph. When it comes to allowing users to build the graph natively (via a pipeline builder UI, for example), we could check for empty `if`s and prevent users from creating them.

Thoughts?

**Comment by original-brownbear** [comment]
@andrewvc for what it's worth, I'm with @ycombinator here. The `Graph` already is a sort of reordered version of the imperative code => I think it's fine to interpret removing redundant statements as being in the same category of reordering and just optimizing them out without ever hitting the `Graph` :)

**Comment by original-brownbear** [comment]
LGTM btw :)

File: logstash-core/src/test/java/org/logstash/config/ir/imperative/IfStatementTest.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+import org.logstash.config.ir.InvalidIRException;
+import org.logstash.config.ir.graph.Graph;
+
+import static org.logstash.config.ir.IRHelpers.*;
+
+public class IfStatementTest {
+
+    @Test
+    public void testEmptyIfStatement() throws InvalidIRException {
```
Comment (andrewvc) [inline logstash-core/src/test/java/org/logstash/config/ir/imperative/IfStatementTest.java]: I realize that this file didn't exist previously. We primarily tested this path through other means.

That said, this feels a bit complete with only the test for the new case we've added.

Would you mind adding one more test for the case where the if statement does have non-noop paths?

Comment (ycombinator) [reply logstash-core/src/test/java/org/logstash/config/ir/imperative/IfStatementTest.java]: No problem at all, happy to add more tests here!

**Comment by ycombinator** [comment]
@andrewvc I've added more unit tests for `IfStatement` in a6c52fc1. This PR is ready for re-review.

**Comment by andrewvc** [comment]
Thanks for adding the comprehensive tests! I apologize if this is too nitpicky, but I do think these tests could be more concise / more accurate. I opened https://github.com/ycombinator/logstash/pull/1 to show how you might do that for one of the tests.

The nice thing about this approach is that since it diffs the whole graph it's a more comprehensive test than just counting things.

It's also shorter to boot.

**Pull request closed without merging**